### PR TITLE
feat(compiler): Add a TSConfig option `useTemplatePipeline`

### PR DESCRIPTION
**Pull request by dylhunn**
The Template Pipeline is a brand new backend for the Angular compiler, replacing `TemplateDefinitionBuilder` and friends. It generates the Ivy instructions corresponding to an input template (or host binding). The Template Pipeline has an all-new design based on an intermediate representation compiled over many phases, which will allow us to experiment with compiler changes more easily in the future.

With this commit, the template pipeline can now be enabled in any project via the `useTemplatePipeline` TSConfig option. However, it is still disabled by default.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts [hunk 1]
```diff
@@ -431,4 +431,9 @@ export interface MiscOptions {
    * another library without option set will not issue error if rendered in orphan way.
    */
   forbidOrphanComponents?: boolean;
+
+  /**
+   * Whether to use TemplateDefinitionBuilder as the code generator, or Template Pipeline.hand
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts:436]: `hand`? ✋

File: packages/compiler/src/jit_compiler_facade.ts [hunk 2]
```diff
@@ -356,7 +357,10 @@ function convertDirectiveFacadeToMetadata(facade: R3DirectiveMetadataFacade): R3
     typeSourceSpan: facade.typeSourceSpan,
     type: wrapReference(facade.type),
     deps: null,
-    host: extractHostBindings(facade.propMetadata, facade.typeSourceSpan, facade.host),
+    host: {
+      ...extractHostBindings(facade.propMetadata, facade.typeSourceSpan, facade.host),
+      useTemplatePipeline: false,
```
Comment (alxhub) [inline packages/compiler/src/jit_compiler_facade.ts:362]: We should probably introduce a constant for these (`SHOULD_USE_TEMPLATE_PIPELINE_FOR_JIT`?) so we don't forget to update them.

**Review by alxhub — APPROVED** [review]

File: packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts (hunk 1 above)
Comment (dylhunn) [reply packages/compiler-cli/src/ngtsc/core/api/src/public_options.ts:436]: I thought this comment might come in _handy_

**Review by mmalerba — APPROVED** [review]

**Comment by dylhunn** [comment]
caretaker: the current two approvals are sufficient to merge.

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 47e6e841016abfca0c1aa84051d82a04b3027617.

**Pull request closed without merging**